### PR TITLE
Bunch of missing `else` statements and general logic

### PR DESCRIPTION
**Pull request by RebelliousX**
Some `else` statements are missing. Also, some of compound statements of the if-else are equal! I don't know if they are intentional.

### CODE REVIEW

**Comment by RebelliousX** [comment]
Please review the changes, I added some comments on some if statements too.

File: main/tests/test_math.cpp
Comment (akien-mga) [inline main/tests/test_math.cpp]: I guess the previous `if` needs an `else` too.

File: modules/visual_script/visual_script_builtin_funcs.cpp
Comment (akien-mga) [inline modules/visual_script/visual_script_builtin_funcs.cpp]: That looks like a bug.

File: drivers/gles3/rasterizer_scene_gles3.cpp
Comment (akien-mga) [inline drivers/gles3/rasterizer_scene_gles3.cpp]: Probably a bug. Would be worth checking the similar code in the gles2 rasterizer if it exists (some code was copy pasted and adapted, could have typos), or the git blame.

File: scene/gui/item_list.cpp
Comment (akien-mga) [inline scene/gui/item_list.cpp]: That actually changes the logic, but I guess the previous behaviour could be considered a bug.

File: servers/physics/shape_sw.cpp
Comment (akien-mga) [inline servers/physics/shape_sw.cpp]: That also changes/fixes a probable logic bug (if `p_normal.z` was positive, the previous case would always be overridden.

File: tools/editor/plugins/texture_region_editor_plugin.cpp
Comment (akien-mga) [inline tools/editor/plugins/texture_region_editor_plugin.cpp]: That's a bug, cc @MarianoGnu.

**Review by akien-mga — APPROVED** [review]
The changes look good to me, the various comments are also relevant and should likely be discussed and fixed before merging.

File: main/tests/test_math.cpp
Comment (RebelliousX) [inline main/tests/test_math.cpp]: Done

**Comment by RebelliousX** [comment]
I added some more, please have a look at them. :)

BTW, I am using PVS-Studio static analyzer **free**. There is no way in hell I could spot these no matter how meticulous would I be.

File: core/variant.cpp
Comment (RebelliousX) [inline core/variant.cpp]: Not sure about this! Does both indices need to be counter `i`? or just one of them? Are we comparing if the two arrays are equal (it seems to me it is)?

Comment (akien-mga) [inline core/variant.cpp]: CC @hpvb.

Comment (hpvb) [inline core/variant.cpp]: Correct, that is broken. Bad @hpvb, bad.

Comment (hpvb) [inline core/variant.cpp]: @RebelliousX Line 2956 has the same bug. I can make a PR for this, or you can roll it in this one.

File: scene/resources/default_theme/default_theme.cpp
Comment (hpvb) [inline scene/resources/default_theme/default_theme.cpp]: I think the bug is elsewhere. With this change the check for p_font.is_valid() becomes useless. If we don't need it remove the whole case.

Comment (akien-mga) [inline scene/resources/default_theme/default_theme.cpp]: No I think the change is fine. As I understand it, the behaviour of this method is (at least it was until now and the change keeps it):
- If p_font is a valid font, use it
- If it isn't, use the default font for either hidpi (if p_hidpi) or lodpi

Comment (akien-mga) [inline scene/resources/default_theme/default_theme.cpp]: And indeed that's the expected logic: https://github.com/godotengine/godot/blob/de0045cf1b0a5e20fbf74da192039d344ee8d0c7/scene/register_scene_types.cpp#L284-L300

Basically the code that calls `make_default_theme` loads a custom font if one is defined, but does not do the validity check, so it has to happen in `make_default_theme` (and if no font is provided or if it's invalid, use the default fonts).

File: core/variant.cpp
Comment (RebelliousX) [inline core/variant.cpp]: @hpvb Done ;)

File: core/io/marshalls.cpp
Comment (akien-mga) [inline core/io/marshalls.cpp]: Is the `ERR_FAIL_COND_V` test useful at all? `count` is a `uint32_t`, so it can't be < 0, right?

File: core/io/stream_peer.cpp
Comment (akien-mga) [inline core/io/stream_peer.cpp]: @Faless An idea maybe?

File: core/pool_allocator.cpp
Comment (akien-mga) [inline core/pool_allocator.cpp]: Then remove that test :)

File: core/variant_parser.cpp
Comment (akien-mga) [inline core/variant_parser.cpp]: I guess the idea is that it shadows the other `token` without modifying its data (while if it modified the `Token&` reference it would change its data). Whether that's necessary or not, I don't know :)

File: modules/regex/regex.cpp
Comment (akien-mga) [inline modules/regex/regex.cpp]: cc @leezh

File: scene/2d/particles_2d.cpp
Comment (akien-mga) [inline scene/2d/particles_2d.cpp]: Hm, why do we have this class rolling its own version of `rand_from_seed` instead of using the Math one?

File: core/io/marshalls.cpp
Comment (akien-mga) [inline core/io/marshalls.cpp]: Alright, that's what your comment meant. Feel free to remove the checks then.

File: core/pool_allocator.cpp
Comment (akien-mga) [inline core/pool_allocator.cpp]: Can be removed.

File: drivers/gles3/shader_compiler_gles3.cpp
Comment (akien-mga) [inline drivers/gles3/shader_compiler_gles3.cpp]: cc @reduz.

File: modules/visual_script/visual_script_builtin_funcs.cpp
Comment (akien-mga) [inline modules/visual_script/visual_script_builtin_funcs.cpp]: cc @reduz.

File: drivers/gles3/rasterizer_scene_gles3.cpp
Comment (akien-mga) [inline drivers/gles3/rasterizer_scene_gles3.cpp]: cc @reduz

File: core/variant_parser.cpp
Comment (akien-mga) [inline core/variant_parser.cpp]: cc @bojidar-bg @reduz

File: scene/gui/rich_text_label.h
Comment (akien-mga) [inline scene/gui/rich_text_label.h]: Sounds buggy indeed.

File: servers/visual/shader_language.cpp
Comment (akien-mga) [inline servers/visual/shader_language.cpp]: cc @reduz

File: tools/editor/import/resource_importer_scene.cpp
Comment (akien-mga) [inline tools/editor/import/resource_importer_scene.cpp]: cc @reduz

File: core/variant.cpp
Comment (hpvb) [inline core/variant.cpp]: This doesn't need sizeof() the macro expands to : 
```c++
_post_initialize(new(_data._mem,sizeof((Dictionary)( p_dictionary)),"") (Dictionary)( p_dictionary))
```

**Comment by akien-mga** [comment]
Nice work overall :) Could you squash the three commits together and edit the commit log to be explicit about the changes?

File: scene/2d/particles_2d.cpp
Comment (hpvb) [inline scene/2d/particles_2d.cpp]: Math::rand_form_seed() uses a different algorithm. Is it possible that this one was implemented to make sure it looks the same when ran on multiple systems, say for networking? This is just a guess of course.

Comment (akien-mga) [inline scene/2d/particles_2d.cpp]: No, it's because it was replaced a first time in a747c8c, and then a second time in 4c90046. I guess both contributors (cc @tagcup) missed that there was code duplication in this class.

File: modules/regex/regex.cpp
Comment (leezh) [inline modules/regex/regex.cpp]: Ahhh, it's supposed to be `7F`. Also, referring to the POSIX standards again, space is excluded too so it should actually be `0x20 < c && c < 0x7F`. 😅

**Comment by RebelliousX** [comment]
@akien-mga I hope I did this right :/ I did interactive rebase and squashed 4 commits.. Is it good? This is the first time doing this :)

File: core/pool_allocator.cpp
Comment (RebelliousX) [inline core/pool_allocator.cpp]: Oh crap! I forgot to remove another one!

File: core/pool_allocator.cpp
Comment (RebelliousX) [inline core/pool_allocator.cpp]: Oh crap! I forgot to remove another one!

**Comment by hpvb** [comment]
@RebelliousX What you want to do is this:
* git rebase -i <id of commit right before your first change>
* then hit 'squash' for all but one of your commits 
* update the commit message to be more useful 

Once that is done when you do 'git log' all your old commits should be gone, instead having just the one commit with your new message. If that is what you see do

git push -f <whatever your own fork's upstream is>

If you're on IRC I can help you with this step by step if you'd like.

**Comment by RebelliousX** [comment]
@hpvb Thanks for the offer and the help. I am using Sourcetree on Windows, which is bogus unlike the Mac version. It kept insisting there was an error, apparently, the force push was disabled on the GUI, so I had to use the terminal in the end which was way easier.

**Comment by akien-mga** [comment]
> so I had to use the terminal in the end which was way easier

For anything git, the command line is always easier. Git is way too powerful and complex to be usable via a GUI past the basic "drag and drop-box" workflow that GitHub promotes :)

File: scene/2d/particles_2d.cpp
Comment (tagcup) [inline scene/2d/particles_2d.cpp]: Oh, sorry. Indeed, I didn't even guess there would be multiple PRNG implementations in the code-base. I'll submit a PR ASAP

File: modules/visual_script/visual_script_builtin_funcs.cpp
Comment (tagcup) [inline modules/visual_script/visual_script_builtin_funcs.cpp]: As far as I can tell, the second one should be `p_idx==1`

File: scene/2d/particles_2d.cpp
Comment (tagcup) [inline scene/2d/particles_2d.cpp]: Submitted PR #7904 for this one

File: modules/visual_script/visual_script_builtin_funcs.cpp
Comment (akien-mga) [inline modules/visual_script/visual_script_builtin_funcs.cpp]: Indeed, it should be `p_idx==1`.

File: scene/2d/particles_2d.cpp
Comment (akien-mga) [inline scene/2d/particles_2d.cpp]: This is now fixed in master, so the addition of the comment should be rebased away as it conflicts.

File: tools/editor/plugins/texture_region_editor_plugin.cpp
Comment (akien-mga) [inline tools/editor/plugins/texture_region_editor_plugin.cpp]: I think you can just remove that empty `else if` so that the second one gets triggered.

File: servers/visual/shader_language.cpp
Comment (akien-mga) [inline servers/visual/shader_language.cpp]: Yes, it should be `||`.

File: core/io/stream_peer.cpp
Comment (akien-mga) [inline core/io/stream_peer.cpp]: It looks like the error handling is actually done in `get_data`, where it returns `ERR_INVALID_PARAMETER` if `p_bytes != r_received`. So maybe this method here should be changed to return `void`, or the error handling should be added directly here and `get_data` should just return `get_partial_data(p_buffer,p_bytes,recv)`.

@Faless @punto- Any idea?

**Comment by akien-mga** [comment]
Looks like SourceTree played one of its magic tricks again, you now have three commits instead of one ;)

**Comment by RebelliousX** [comment]
@akien-mga Actually, I used github's resolve conflict feature. I didn't touch Sourcetree till now, but now, it even refuses to do rebase because github merged branches I think :/

File: core/io/stream_peer.cpp
Comment (Faless) [inline core/io/stream_peer.cpp]: I've never used the `StreamPeerBuffer` class myself but judging from the commit log of cbbcf727035c8b481889f605337a96a9e58ed970 and it's GDScript signature it looks cool.

`get_partial_data` is inherited from `PacketPeer` and we cannot change it's type.
It also appears to be exposed to GDScript `StreamPeerBuffer.get_partial_data(...)` via `StreamPeer::_get_partial_data(...)`.
It think this should return a proper error when needed, so option 2, something like
```
diff --git a/core/io/stream_peer.cpp b/core/io/stream_peer.cpp
index a64ebf2..bf6a861 100644
--- a/core/io/stream_peer.cpp
+++ b/core/io/stream_peer.cpp
@@ -479,12 +479,15 @@ Error StreamPeerBuffer::get_data(uint8_t* p_buffer, int p_bytes){
 }
 Error StreamPeerBuffer::get_partial_data(uint8_t* p_buffer, int p_bytes,int &r_received){
 
+       Error err = OK;
 
        if (pointer+p_bytes > data.size()) {
+               err = ERR_FILE_EOF;
+
                r_received=data.size()-pointer;
                if (r_received<=0) {
                        r_received=0;
-                       return OK; //you got 0
+                       return err; //you got 0
                }
        } else {
                r_received=p_bytes;
@@ -492,6 +495,8 @@ Error StreamPeerBuffer::get_partial_data(uint8_t* p_buffer, int p_bytes,int &r_r
 
        PoolVector<uint8_t>::Read r = data.read();
        copymem(p_buffer,r.ptr(),r_received);
+
+       return err;
 }
 
 int StreamPeerBuffer::get_available_bytes() const {
```

**Comment by akien-mga** [comment]
Let's merge the current state already as it's pretty good. I'll make sure the remaining FIXMEs get fixed soon too.

**Comment by tagcup** [comment]
@akien-mga @RebelliousX There are two bugs left unmarked (meaning no recognizable labels such as "FIXME"):

```
else if (p_idx==0)	// is it ok to test p_idx == 0 twice?
```

in `modules/visual_script/visual_script_builtin_funcs.cpp` and

```
		/* Why are both statements equal? */
  		storage->shaders.copy.set_conditional(CopyShaderGLES3::LINEAR_TO_SRGB,true);
```

in `drivers/gles3/rasterizer_scene_gles3.cpp`.
```

**Pull request merged**
